[WIP]: Use `react-query` for data fetching/mutations

sam-potter opened the pull request:
# Work in progress. Do not merge.

## Intro

This pull request is a continuation of my previous PR (#577) where I introduced `react-query` for fetching and caching the treasury dashboard statistics. I want to continue pushing `react-query` as a server-state management solution by slowly converting most of the redux slices. Hopefully, this will show just how simple it makes handling asynchronous logic, whilst also avoiding infinitely many bugs.

Would love to hear comments/criticisms. Would also love to get onboarded to the Discord and chat further on there.

## Issues

### Standarized network enum

I thought the first place to start would be around how the app handles the currently active network in the user's wallet. The first step with tackling some like this should always be creating a proper type to describe the data.

I extracted the `NetworkID` enum in `src/lib/Bond.ts` to a new network constants file (`src/constants/networks.ts`) and made several alterations.

```jsx
// Using numbers in enums like this can be problematic.
export enum NetworkID {
  Mainnet = 1,
  Testnet = 4,
  Arbitrum = 42161,
  ArbitrumTestnet = 421611,
  AvalancheTestnet = 43113,
  Avalanche = 43114,
}

// Because while it is intended to be used like this
const networkId: NetworkID = NetworkID.Mainnet;

// Stuff like this is actually also valid too.
const networkId: NetworkID = 2;
const networkId: NetworkID = Math.random() / 10e8 * 0.412341;

// It's the same reason why the initial state in NetworkSlice.ts doesn't throw an error when it should.
const initialState: INetworkSlice = {
  networkId: -1,
  networkName: "",
  uri: "",
  initialized: false,
};
```

As such, the new type looks like this

```jsx
export enum NetworkId {
  MAINNET = "1",
  TESTNET_RINKEBY = "4",

  ARBITRUM_ONE = "42161",
  ARBITRUM_TESTNET_RINKEBY = "421611",

  AVALANCHE = "43114",
  AVALANCHE_TESTNET_FUJI = "43113",
}
```

We can convert all the numbers to strings (because both primitives index objects similarly) so the enum no longer suffers from the earlier issues. Moreover, the names of each network are capitalised since they are effectively constants, whilst also having more standardised and accurate names. The enum is also properly `PascalCased`.

### Network information queries/mutations

I think there are a couple issues with how data is implemented across the codebase at the moment:

#### 1. A lot of data fetching isn't "reactive". 
There are too many instances, especially in the main `App.tsx` component, where data is fetched on command. What that ends up creating is a web of specifying when certain data in the store should be loaded and refreshed across too many different components. As someone just beginning to contribute to this codebase, it was a huge hurdle to try and understand where all the data is coming from. Data should be able to automatically work out when it's needed.

#### 2. Too much code duplication. 
Nearly all of the slices take in multiple arguments that are almost identical, irrelevant of the specific data being fetched. These arguments mean whenever a `dispatch` is fired from a component, that component must first house several hooks that fetch data that is only used as an argument for more data.

#### 3. Incomplete state machines. 
Handling asynchronous operations means we need to handle multiple states while data is pending, being refreshed, or an error occurred. It's a hassle to do, and very difficult to get right.

You can see how some of these problems are addressed in the first few hooks I wrote in `src/hooks/useNetwork`. Feel free to look at that file side-by-side with the `NetworkSlice.ts` file, since the two files almost functionality identical. All the function arguments can be omitted and replaced with the actual hooks that provide that info. Furthermore, all the states that surround fetching/mutating that data are backed by the finite state machines built into the `react-query` hooks with correct typings, ensuring that if a `query.status === 'success'`, Typescript recognises that the data is guaranteed to be there.

Keep in mind, this PR is a work in progress, and would really appreciate any criticisms and thoughts to ensure that this upgrade is worth the time, and done in the best way possible.

sam-potter committed:
Elevate queryProvider

greg-schrammel commented on src/constants.ts:
```diff
@@ -12,21 +8,13 @@ export const BLOCK_RATE_SECONDS = 13.14;
-
-export const addresses: IAddresses = {
-  4: {
+export const addresses: Record<NetworkId, Record<string, string>> = {
```
if we just remove the typing ts can infer and autocomplete the addresses, but it will throw on some places where there are no address for the network being passed, don't know if it's worth making the change rn
![image](https://user-images.githubusercontent.com/6232729/147239822-f22f977c-ea50-42e5-8bd2-f8c1ef5a94e8.png)

sam-potter replied:
Agreed, if inference allows us to use autocomplete I think it's worth removing the explicit types.

Won't change it right now though, am still playing around to find possibly better ways of handling contracts and their addresses.

sam-potter committed:
Refactored provider + network logic

sam-potter committed:
Infer provider type

sam-potter committed:
Remove redundant query client

sam-potter committed:
Added first few hooks

sam-potter commented:
With #1012 merged, I've gone ahead and rebased this branch on top of `develop` and made the necessary changes to use the new enum.

sam-potter committed:
Set up core elements for new provider logic

sam-potter committed:
Temporarily remove provider connection analytics

sam-potter committed:
Add queries for wallet address + ens info

sam-potter committed:
Added new contract functionality

sam-potter committed:
Added valid address invariant

sam-potter committed:
Network now in-sync when wallet connected

sam-potter committed:
Added support for contracts from explicit networks

sam-potter committed:
Implemented all queries for the dashboard metrics

sam-potter commented:
## Progress update

I've just finished implementing the first actual changes to the UI - the metrics on the dashboard now fetch their data entirely through `react-query` - so I thought it would be a good time to give an update on the status of this PR.

### New provider logic

1. There are now only ever two providers available over the lifetime of the app: a `StaticJsonRpcProvider` and a `Web3Provider`. On initial load, we use the static provider that is connected to a third-party service like Infura, Alchemy, etc. Once a wallet is connected, we use the provider injected into the browser by that wallet (or the wallet's third party service).
2. `useEagerConnect` is the new hook that checks if a user has connected a wallet before, and connects to it if they have.
3. `useProviderEventListeners` handles all necessary event listeners for the wallet (events like changing accounts, changing networks, etc) and updates necessary data accordingly.
4. `useNetwork` is the new hook responsible for handling the currently active network and defaults to mainnet.

### New contract retrieval strategy

The way contracts are currently retrieved is from a list of all contracts available for a given network. The new way is effectively the opposite: a list of all networks available for a given contract.

This makes it easier when handling contracts across changing networks: we simply pass in a list of the addresses for a contract on all the networks it exists and the new `src/hooks/useContract` hook will return that contract on the currently active network (or null if that specific contract doesn't exist for the currently active network).

However, sometimes we only want a contract from a specific network (and we don't care about the currently active network), that is also supported too with all the necessary function overloads and assertions to ensure that contract actually exists. You can see this in action in `src/hooks/useCurrentIndex` and `src/hooks/useMarketPrice` where the contracts returned from the hooks are never null since we are using mainnet only contracts for this info.

### Dashboard metrics

All this new stuff has been implemented on the dashboard page. I recommend you checkout to this PR and run it yourself locally to visibly see the improvement: you'll see that the metrics load faster, especially when no wallet is connected. Also check out the code for it in `Metric.tsx` to see just how simple the data fetching hooks are, whilst also causing far fewer re-renders.

sam-potter committed:
Improved type safety

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

sam-potter committed:
Converted `Metric.js` to `Metric.tsx`

sam-potter committed:
Refactor

sam-potter committed:
Extracted `useGohmPrice` hook

sam-potter committed:
Refactored token price hooks

sam-potter committed:
Removed block number from graph query

sam-potter committed:
Improved `Object.keys` typings

sam-potter committed:
Added more contract addresses

sam-potter committed:
Refactor

sam-potter committed:
Added covalent API interface + types

sam-potter committed:
Added hooks for wallet token balances

sam-potter commented:
Just added the hooks for fetching wallet balances.

Because of the new way contract addresses are handled, we get automatic typings about which networks our token exists on when getting it's balance!

![Screen Shot 2022-01-08 at 6 18 30 pm](https://user-images.githubusercontent.com/15123145/148635836-fbf92501-0aae-4bf3-bc6a-827cc5f25782.png)
 
And of course warns us when trying to get its balance on a network where it doesn't exist.

![Screen Shot 2022-01-08 at 6 20 24 pm](https://user-images.githubusercontent.com/15123145/148635875-548750fd-4c97-4f7a-995c-3e0ca1d33e45.png)

sam-potter committed:
Refactor

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

sam-potter committed:
Use v2 staking interface

sam-potter committed:
Prefer BigNumber over number

sam-potter committed:
Added BigNumber utility functions

sam-potter committed:
Added pending transactions to covalent API

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

sam-potter committed:
Revert "Merge branch 'develop' into sam-potter/react-query-implementation"

This reverts commit 1b1510e7da205bc554346b057f7c825b2b473a37.

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

sam-potter committed:
Added hook for contract allowances

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

nabioz commented on src/hooks/usePrices.ts:
```diff
@@ -0,0 +1,46 @@
+  return useQuery<number, Error>(ohmPriceQueryKey(), async () => {
+    const [ohm, dai] = await reserveContract.getReserves();
+
+    return parseFloat(formatUnits(dai.div(ohm), 9));
```
It would be nice if we put the decimal magic numbers in a constants file to know how they were determined

sam-potter replied:
Done

nabioz commented on src/hooks/usePrices.ts:
```diff
@@ -0,0 +1,46 @@
+    async () => {
+      queryAssertion(ohmPrice && currentIndex, gohmPriceQueryKey(ohmPrice, currentIndex));
+
+      return parseFloat(formatUnits(currentIndex, 9)) * ohmPrice;
```
Same here with the decimal number

sam-potter replied:
Done

nabioz commented on src/constants/addresses.ts:
```diff
@@ -0,0 +1,47 @@
+  [NetworkId.ARBITRUM]: "0x8D9bA570D6cb60C7e3e0F31343Efe75AB8E65FB1",
+  [NetworkId.AVALANCHE]: "0x321e7092a180bb43555132ec53aaa65a5bf84251",
+  [NetworkId.POLYGON]: "0xd8cA34fd379d9ca3C6Ee3b3905678320F5b45195",
+  [NetworkId.FANTOM]: "0xd8cA34fd379d9ca3C6Ee3b3905678320F5b45195",
```
Fantom address here is wrong

sam-potter replied:
Fixed

nabioz commented on src/helpers/index.tsx:
```diff
@@ -306,6 +308,53 @@ export const handleContractError = (e: any) => {
+ * @returns Formatted string representation of OHM equivalent.
+ */
+export const convertGohmToOhm = (amount: BigNumber, index: BigNumber): string => {
+  return formatUnits(amount.div(10 ** 9).mul(index), 36);
```
Can refactor these magic numbers as constants

appleseed-iii replied:
agree

sam-potter replied:
I'll be honest here, these two decimals numbers in the `convertGohmToOhm` and `convertOhmToGohm` functions are magical. I'm not really sure why it needs to be 36 and 18 respectively, just that those are the correct values to put the decimal in the right spot. I'll try look into why that's the case tomorrow and update you with what I find.

sam-potter replied:
Essentially, it's do with how BigNumber handles math operations internally. It's weird here because we are converting between gohm (an 18 decimal number) and ohm (only a 9 decimal number), so it adds padding to ensure we don't lose precision when dividing/multiplying between the two.

sam-potter replied:
Is it alrght if I leave these specific two numbers as is? I feel like it doesn't really make sense to make these their own constants, but up to you guys @mrunbanked @appleseed-iii

appleseed-iii replied:
yea I think that's fine

nabioz reviewed (changes requested):
Overall I like where this is going. I'm new to the react-query setup but this PR looks harmless to me and might be nice to merge and then get a feel of the developer experience with this architecture.

nabioz commented:
![image](https://user-images.githubusercontent.com/92614968/151173064-a3b589a7-b772-46fe-af6d-b0b2314e944a.png)
Connecting the wallet is not refreshing the UI and the modal just gets stuck there

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

sam-potter committed:
Removed new web3context

sam-potter committed:
Added static  argument

sam-potter committed:
Fixed incorrect fantom address

sam-potter committed:
Fixed failing tests

appleseed-iii commented on src/Root.tsx:
```diff
@@ -17,13 +20,17 @@ const Root: FC = () => {
-        </I18nProvider>
-      </Provider>
+      <NewWeb3ContextProvider>
+        <QueryClientProvider client={queryClient}>
```
can you explain the logic in moving QueryClientProvider up here?

sam-potter replied:
Just because fewer re-renders occur in the Root component compared to the App component at the moment, and re-rendering the react-query provider is pointless, so it'll be a slight perf buff.

appleseed-iii replied:
nice, thank you ser

appleseed-iii commented on src/helpers/index.tsx:
```diff
@@ -306,6 +308,53 @@ export const handleContractError = (e: any) => {
+/**
+ * Converts a supported networkId to its hexadecimal value as a string
+ */
+export const networkIdToHexString = (id: NetworkId) => "0x" + id.toString(16);
```
we already have `idToHexString` in `helpers/NetworkHelper.ts`

sam-potter replied:
Should I remove this one and just use the NetworkHelper version then? Only put it here because it seemed like an appropriate place for a generic helper function like this

appleseed-iii replied:
IMO yea I would remove this one & use that one just so we don't repeat the code two ways.

sam-potter replied:
Done

appleseed-iii commented on src/helpers/index.tsx:
```diff
@@ -306,6 +308,53 @@ export const handleContractError = (e: any) => {
+/**
+ * Determines if app is viewed within an <iframe></iframe>
+ */
+export const isIFrame = () => window.location !== window.parent.location;
```
we have an implementation of this in `web3Context`. Can you remove the one in web3Context?

appleseed-iii replied:
scratch that, let's leave that as a TODO in a later pr to keep this one smaller.

sam-potter replied:
Agreed, later PR it is

appleseed-iii commented on src/hooks/useSwitchNetwork.ts:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Switches the wallets currently active network.
+ */
+export const useSwitchNetwork = () => {
```
can this replace `export const switchNetwork` in `src/helpers/NetworkHelper`

sam-potter replied:
Yep absolutely, once this hook is actually being used in the app i'll then remove `switchNetwork` from `src/helpers/NetworkHelper`.

sam-potter committed:
Added contract decimals constants

appleseed-iii commented on src/hooks/useBalances.ts:
```diff
@@ -0,0 +1,86 @@
+  });
+};
+
+export const useOhmBalance = () => useBalance(OHM_ADDRESSES);
+export const useSohmBalance = () => useBalance(SOHM_ADDRESSES);
+export const useGohmBalance = () => useBalance(GOHM_ADDRESSES);
+export const useWsohmBalance = () => useBalance(WSOHM_ADDRESSES);
+export const useV1OhmBalance = () => useBalance(V1_OHM_ADDRESSES);
+export const useV1SohmBalance = () => useBalance(V1_SOHM_ADDRESSES);
```
are these not implemented yet?

sam-potter replied:
Correct, not actually used anywhere in the UI atm. Just wrote them to show how to compose the `useBalance` hook into smaller, more specific hooks.

sam-potter commented:
@mrunbanked 



> ![image](https://user-images.githubusercontent.com/92614968/151173064-a3b589a7-b772-46fe-af6d-b0b2314e944a.png) Connecting the wallet is not refreshing the UI and the modal just gets stuck there

Fixed now

sam-potter commented:
I've gone ahead and removed the new web3context, as well as the `useAddress` and `useNetwork` react-query hooks, in favour of the existing web3context. Couple reasons why:

1. Using react-query as the source for that info didn't really add any benefits and actually just complicated it more.
2. The slight improvements that I had implemented in the new provider we're only half way there, I felt like changing anything further was creeping outside the scope of this PR.

I'll open up a new PR re all of this over the next few days to discuss this further.

sam-potter committed:
Merge branch 'develop' into sam-potter/react-query-implementation

appleseed-iii reviewed (approved):
approved pending that one minor comment about networkid hex

sam-potter committed:
Remove networkIdToHexString
